### PR TITLE
Report L1 base fee set by L2s

### PR DESCRIPTION
**Pull request by matYang**
## Motivation
On optimistic rollups, a major portion of tx cost is data availability fee (DA fee). DA fee = DA gas * L1 gas price.
DA gas can be calculated from message's calldata size.
L1 gas price needs to be read from the L2's L1 gas oracle.

## Solution
Read L2-specific L1 gas price, encode it to PriceRegistry-compatible int format, where higher-order 112 bits are L1 gas price, and lower-order 112 bits are exec gas price.

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/execution_gas_helpers.go [hunk 1]
```diff
@@ -64,9 +66,29 @@ func maxGasOverHeadGas(numMsgs, dataLength, numTokens int) uint64 {
...
+	if execGasPriceEstimate.DynamicFeeCap != nil {
+		destGasPrice = execGasPriceEstimate.DynamicFeeCap.ToInt()
+	}
+
+	execGasEstimate := new(big.Int).Add(big.NewInt(FEE_BOOSTING_OVERHEAD_GAS), msg.GasLimit)
+	execGasEstimate = new(big.Int).Mul(execGasEstimate, destGasPrice)
+
+	if execGasPriceEstimate.L1BaseFee != nil {
+		if l1BaseFee := execGasPriceEstimate.L1BaseFee.ToInt(); l1BaseFee.Cmp(big.NewInt(0)) > 0 {
+			MESSAGE_FIXED_BYTES := 448
```
Comment (matYang) [inline core/services/ocr2/plugins/ccip/execution_gas_helpers.go:80]: quick and dirty examples, don't mind these magic numbers for now

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 2]
```diff
@@ -572,12 +596,12 @@ func calculateIntervalConsensus(intervals []commit_store.CommitStoreInterval, f
...
 // The provided latestTokenPrices should not contain nil values.
-func (r *CommitReportingPlugin) calculatePriceUpdates(observations []CommitObservation, latestGasPrice update, latestTokenPrices map[common.Address]update) commit_store.InternalPriceUpdates {
+func (r *CommitReportingPlugin) calculatePriceUpdates(observations []CommitObservation, latestGasPrice gasPriceUpdate, latestTokenPrices map[common.Address]tokenPriceUpdate) commit_store.InternalPriceUpdates {
 	priceObservations := make(map[common.Address][]*big.Int)
-	var sourceGasObservations []*big.Int
+	var sourceGasObservations []GasPrice
 
 	for _, obs := range observations {
-		if obs.SourceGasPriceUSD != nil {
+		if obs.SourceGasPriceUSD.notNil() {
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: aside: we should move this validation into getParseableObservations and we should similarly move the > r.F check into the top level report instead of spread out in 3 places. Just consider observations all or nothing

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 3]
```diff
@@ -339,17 +345,35 @@ func (r *CommitReportingPlugin) generatePriceUpdates(
...
+		return GasPrice{}, nil, fmt.Errorf("missing gas price %+v", sourceGasPriceWei)
+	}
+
+	sourceNativeGasPriceUSD := calculateUsdPerUnitGas(gasPrice, sourceNativePriceUSD)
+	sourceGasPriceUSD = GasPrice{
+		DAGasPrice:     big.NewInt(0),
+		NativeGasPrice: sourceNativeGasPriceUSD,
+	}
+
+	// If l1 oracle exists, l1 gas price is a price component of overall tx, need to fetch and encode l1 gas price.
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: seems very hard to follow what will happen with different versions here and the plugin doesn't need to know any of these details. I think we can abstract all of this into 2 helper components (which wrap gas.EvmGasEstimator) one for 1.0/1.1 and one for 1.2 which implement a common interface that the plugin uses. Something like:
```
type GasPriceEstimator interface {
	EstimateGasPrice() (GasPrice, error) // plugin calls this then EncodeGasPrice in Observe
	EncodeGasPrice(GasPrice) ([]byte, error) // encodes according to version
	DecodeGasPrice([]byte)(GasPrice, error) //  decodes according to version
}

type GasPrice interface {
	String() string // for logging
	Cmp(GasPrice) int // Used for deviation and median
}
```
When we instantiate the plugin:
```
var gasPriceEstimator GasPriceEstimator
switch commitVersion 
case 1_0, 1_1 {
    gasPriceEstimator = NewV1GasPriceEstimator() // native fees only
} case 1_2 {
    gasPriceEstimator = NewV2GasPriceEstimator() // native fees and da
} default {
	error
}
commitPlugin{gasPriceEstimator: gasPriceEstimator...}
```
Has an added benefit of improving testability, smaller interface injected vs the whole gas.EvmFeeEstimator and it will help reduce duplicate code in the exec plugin I think

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 4]
```diff
@@ -407,18 +431,18 @@ func (r *CommitReportingPlugin) getLatestTokenPriceUpdates(ctx context.Context,
...
 		latestInflightGasPriceUpdate := r.inflightReports.getLatestInflightGasPriceUpdate()
-		if latestInflightGasPriceUpdate != nil && latestInflightGasPriceUpdate.timestamp.After(gasPriceUpdate.timestamp) {
-			gasPriceUpdate = *latestInflightGasPriceUpdate
+		if latestInflightGasPriceUpdate != nil && latestInflightGasPriceUpdate.timestamp.After(gasUpdate.timestamp) {
+			gasUpdate = *latestInflightGasPriceUpdate
 		}
 
-		if gasPriceUpdate.value != nil {
-			r.lggr.Infow("Latest gas price from inflight", "gasPriceUpdateVal", gasPriceUpdate.value, "gasPriceUpdateTs", gasPriceUpdate.timestamp)
+		if gasUpdate.value.notNil() {
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: I think clarity would be improved if we just explicitly return a boolean indicating whether there is a prior gas price update and remove this "notNil means no update"

File: core/services/ocr2/plugins/ccip/commit_plugin.go [hunk 5]
```diff
@@ -66,6 +67,14 @@ func NewCommitServices(lggr logger.Logger, jb job.Job, chainSet evm.LegacyChainC
 	if err != nil {
 		return nil, errors.Wrap(err, "failed getting the static config from the commitStore")
 	}
+	typeAndVersion, err := commitStore.TypeAndVersion(&bind.CallOpts{})
```
Comment (RensR) [inline core/services/ocr2/plugins/ccip/commit_plugin.go]: Maybe we should change the loaders to also return the version, as we already parse it in `LoadCommitStore`

File: core/services/ocr2/plugins/ccip/config/offchain_config.go [hunk 6]
```diff
@@ -15,6 +15,50 @@ type OffchainConfig interface {
...
+		return errors.New("must set MaxGasPrice")
+	}
+	if c.InflightCacheExpiry.Duration() == 0 {
+		return errors.New("must set InflightCacheExpiry")
+	}
+
+	return nil
+}
+
+type CommitOffchainConfigV1 struct {
```
Comment (RensR) [inline core/services/ocr2/plugins/ccip/config/offchain_config.go:62]: nit: add a comment to indicate the version range this is valid for

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 2 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: seems tricky, since `interval | token price | gas price` are validated individually, e.g. an Observation where gas price is nil can contain valid values for token price and/or interval, I think it'd still make sense to check them separately

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 4 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: Can change getLatestInflightGasPriceUpdate to return (update, bool) as opposed to *update.
Per my understanding a priori gas price update where chain selector is not 0 should never contain a gas value that's nil, was this nil check more intended as defensive code?

Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: Checks removed

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 2 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: Communicated on Slack, decision is to reject Observations if any price field is nil, since with current impl, non-faulty nodes should not produce observations with nil fields.

File: core/services/ocr2/plugins/ccip/commit_inflight.go [hunk 7]
```diff
@@ -67,28 +67,30 @@ func (c *inflightCommitReportsContainer) maxInflightSeqNr() uint64 {
...
+	updateFound := false
+	latestGasPriceUpdate := update{}
 	var latestEpochAndRound uint64
 	for _, inflight := range c.inFlightPriceUpdates {
 		if inflight.priceUpdates.DestChainSelector == 0 {
 			// Price updates did not include a gas price
 			continue
 		}
-		if latestGasPriceUpdate == nil || inflight.epochAndRound > latestEpochAndRound {
+		if !updateFound || inflight.epochAndRound > latestEpochAndRound {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/commit_inflight.go:82]: You can stop the iteration when the first update is found.

```
// that's what we are doing now
updateFound := false

for _, inflight := range c.inFlightPriceUpdates {
    if !updateFound {
       updateFound = true
    }
}
```

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 8]
```diff
@@ -402,19 +418,18 @@ func (r *CommitReportingPlugin) getLatestTokenPriceUpdates(ctx context.Context,
 	return latestUpdates, nil
 }
 
-// Gets the latest gas price updates based on logs within the heartbeat
-func (r *CommitReportingPlugin) getLatestGasPriceUpdate(ctx context.Context, now time.Time, checkInflight bool) (gasPriceUpdate update, error error) {
+// getLatestGasPriceUpdate returns the latest gas price updates based on logs within the heartbeat.
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: typo:  `gas price update based`

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 9]
```diff
@@ -499,6 +521,54 @@ func (r *CommitReportingPlugin) Report(ctx context.Context, epochAndRound types.
 	return true, encodedReport, nil
 }
 
+// validateObservations validates the given observations.
+// An observation is rejected if any of its gas price or token price is nil. With current CommitObservation implementation, prices
+// are checked to ensure no nil values before adding to Observation, hence an observation that contains nil values comes from a faulty node.
+func (r *CommitReportingPlugin) validateObservations(ctx context.Context, lggr logger.Logger, observations []CommitObservation) (validObs []CommitObservation, err error) {
+	tokenDecimals, err := r.tokenDecimalsCache.Get(ctx)
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: I think it's somewhat better to read from token decimals cache in the caller and pass the map to this function, that way it's easier to test and easier to re-use the map.

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 10]
```diff
@@ -499,6 +521,54 @@ func (r *CommitReportingPlugin) Report(ctx context.Context, epochAndRound types.
...
+	}
+
+	for _, obs := range observations {
+		// If gas price is reported as nil, the observation is faulty, skip the observation.
+		if obs.SourceGasPriceUSD == nil {
+			lggr.Warnw("Skipping observation due to nil SourceGasPriceUSD")
+			continue
+		}
+		// If observed number of token prices does not match number of supported tokens on dest chain, skip the observation.
+		if len(tokenDecimals) != len(obs.TokenPricesUSD) {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: While this check is logically correct I feel it causes confusion.
We shouldn't mix the `tokenDecimals` map with the supported tokens on dest chain
better to do something like this

```
supportedDestTokens = tokendecimalsMap.Keys()

validateObservations(supportedTokens, ...)
```

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 11]
```diff
@@ -499,6 +521,54 @@ func (r *CommitReportingPlugin) Report(ctx context.Context, epochAndRound types.
...
+			lggr.Warnw("Skipping observation due to token count mismatch", "expecting", len(tokenDecimals), "got", len(obs.TokenPricesUSD))
+			continue
+		}
+		// If any of the observed token prices is reported as nil, or not supported on dest chain, skip the observation.
+		skipObservation := false
+		for token, price := range obs.TokenPricesUSD {
+			if price == nil {
+				lggr.Warnw("Nil value in observed TokenPricesUSD", "token", token.Hex())
+				skipObservation = true
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go:548]: Maybe it's safe to stop on the first skipped observation?

i.e. with something like this
```
func skipObservation() bool {
   for token, price := range obs.TokenPricesUSD {
      if price==nil {
           return false
      }

      if _, exists := tokenDecimals[token]; !exists {
          return false
      }
   }
   return true
}
```

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin_test.go [hunk 12]
```diff
@@ -411,6 +458,156 @@ func TestCommitReportingPlugin_ShouldTransmitAcceptedReport(t *testing.T) {
...
+		},
+		{
+			name:               "too many faulty observations with f=1",
+			commitObservations: []CommitObservation{ob1, nilOb},
+			f:                  1,
+			expValidObs:        nil,
+			expError:           true,
+		},
+		{
+			name:               "all faulty observations",
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin_test.go:596]: maybe include a test case where the observation is skipped because the token does not exist in the `tokenDecimals` map

File: core/services/ocr2/plugins/ccip/config/offchain_config.go [hunk 13]
```diff
@@ -15,6 +15,51 @@ type OffchainConfig interface {
...
+	GasPriceHeartBeat        models.Duration
+	DAGasPriceDeviationPPB   uint32
+	ExecGasPriceDeviationPPB uint32
+	TokenPriceHeartBeat      models.Duration
+	TokenPriceDeviationPPB   uint32
+	MaxGasPrice              uint64
+	InflightCacheExpiry      models.Duration
+}
+
+func (c CommitOffchainConfig) Validate() error {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/config/offchain_config.go:29]: there is a very nice plugin that is already in our go mod
```
https://github.com/go-playground/validator/
```

if we use it here then it's going to be something like this:

```
type CommitOffChainConfig struct {
   SourceFinalityDepth uint32 `validate:not-zero`,
   ...
}

func (c CommitOffchainConfig) Validate() error {
   return govalidate.validate(c)
}
```

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 14]
```diff
@@ -63,13 +64,15 @@ type ExecutionPluginConfig struct {
 	onRamp                   evm_2_evm_onramp.EVM2EVMOnRampInterface
 	offRamp                  evm_2_evm_offramp.EVM2EVMOffRampInterface
 	commitStore              commit_store.CommitStoreInterface
+	commitStoreVersion       semver.Version
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:68]: hmm I don't really like the idea of having a struct from an external library as part of our plugin config.

Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:68]: we could create our own type

Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:68]: or even a small lib to parse semver

File: core/services/ocr2/plugins/ccip/prices/da_price_estimator.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+func (g DAGasPriceEstimator) GetGasPrice(ctx context.Context) (GasPrice, error) {
+	execGasPrice, err := g.execEstimator.GetGasPrice(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var gasPrice *big.Int = execGasPrice
+	if gasPrice.BitLen() > int(g.priceEncodingLength) {
+		return nil, fmt.Errorf("native gas price exceeded max range %+v", gasPrice)
+	}
+
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/prices/da_price_estimator.go:32]: you could simplify the code by doing:

```
if g.l1Oracle == nil {
   return execGasPrice
}
```

File: core/services/ocr2/plugins/ccip/commit_inflight.go (hunk 7 above)
Comment (connorwstein) [reply core/services/ocr2/plugins/ccip/commit_inflight.go:82]: I don't think he can since we still need to look for a potentially later epochAndRound in the remaining inflight reports? Might be cleaner to just split price and gas updates when we `add` but can be a follow up, then you could use sort() and len()=0

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go (hunk 14 above)
Comment (connorwstein) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:68]: this will go away once I abstract the commit store, so very transient

File: core/services/ocr2/plugins/ccip/prices/exec_price_estimator.go [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (g ExecGasPriceEstimator) DenoteInUSD(p GasPrice, wrappedNativePrice *big.Int) (GasPrice, error) {
+	return ccipcalc.CalculateUsdPerUnitGas(p, wrappedNativePrice), nil
+}
+
+func (g ExecGasPriceEstimator) Median(gasPrices []GasPrice) (GasPrice, error) {
+	var prices []*big.Int
+	for _, p := range gasPrices {
+		if p != nil {
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/prices/exec_price_estimator.go]: the rest of the implementations assume non-nil input. I think we either make the input trusted to be not-nil (i.e. document that in the interface) and don't check it or we check it in all the impls and return an error. Probably the former since we validate input in the plugin

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 11 above)
Comment (connorwstein) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go:548]: yeah agreed I think we can just continue on first invalidation

File: core/services/ocr2/plugins/ccip/prices/gas_price_estimator.go [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+	// Use this in per-token gas cost calc as heuristic to simplify estimation logic.
+	ExecGasPerToken = 10_000
+	// ExecGasPerPayloadByte is gas charged for passing each byte of `data` payload to CCIP receiver, ignores 4 gas per 0-byte rule.
+	// This can be a constant as it is part of EVM spec. Changes should be rare.
+	ExecGasPerPayloadByte = 16
+	// EvmMessageFixedBytes is byte size of fixed-size fields in EVM2EVMMessage
+	// Updating EVM2EVMMessage involves an offchain upgrade, safe to keep this as constant in code.
+	EvmMessageFixedBytes     = 448
+	EvmMessageBytesPerToken  = 128          // Byte size of each token transfer, consisting of 1 EVMTokenAmount and 1 bytes, excl length of bytes
+	DAMultiplierBase         = int64(10000) // DA multiplier is in multiples of 0.0001, i.e. 1/DAMultiplierBase
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/prices/gas_price_estimator.go]: shouldn't need to expose these version specific details from the prices package?

File: core/services/ocr2/plugins/ccip/commit_inflight.go (hunk 7 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_inflight.go:82]: yup, `add` and `expire` of the cache operate on `createdAt`, so the cache is implicitly sorted in insertion order, per my understanding insertion order into the cache does not guarantee ascending order of `epochAndRound`, as Connor suggested this could be a followup improvement to the cache.

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 10 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: The mapping is perfect for checking if token in obs exists, converting it to slice before passing makes things tad bit more difficult. For clarity I renamed it to `supportedTokensMap` and added comment, let me know wdyt.

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 11 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go:548]: This was on purpose, as nil prices should not happen - it indicates a bug or a faulty node, normally this should never trigger, when it does, printing all nil values will help with triaging.

File: core/services/ocr2/plugins/ccip/config/offchain_config.go (hunk 13 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/config/offchain_config.go:29]: good find; looks like the package is not being used in the repo today tho, and need to add a custom `ValidateValuer` for `model.Duration`, let's not expand scope of this PR further.

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go (hunk 14 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:68]: will be removed in compatibility phase 2 =)

File: core/services/ocr2/plugins/ccip/commit_inflight.go (hunk 7 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/commit_inflight.go:82]: oh nvm, it's an `OR` I miss-read this

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 10 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go]: Still a bit tricky because the map is `address -> uint8` but it is supposed to contain only the supported tokens.
But should be ok, it's not in the scope of this PR.

Actually the problem starts in the `Report` method where we assume that the supported tokens are the tokens that exist in the decimals cache.
```
supportedTokensMap, err := r.tokenDecimalsCache.Get(ctx)
```

I will clean this up as a follow up in another PR.

File: docs/CHANGELOG_CCIP.md [hunk 18]
```diff
@@ -20,6 +20,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
     - `destDataAvailabilityOverheadGas` is the extra data availability gas charged on top of message data.
     - `destGasPerDataAvailabilityByte` is the amount of gas to charge per byte of data that needs data availability.
     - `destDataAvailabilityMultiplier` is the multiplier for data availability gas. It is in multiples of 1e-4, or 0.0001. It can represent calldata compression factor on Optimistic Rollups.
+- MessageId hashing logic updated.
+  - the new `sourceTokenData` field is added to the hash.
+  - fixed-size fields are hashed in nested hash function.
+- CommitStore OffchainConfig fields updated.
+  - New fields `GasPriceHeartBeat`, `DAGasPriceDeviationPPB`, `ExecGasPriceDeviationPPB`, `TokenPriceHeartBeat`, `TokenPriceDeviationPPB` added
```
Comment (connorwstein) [inline docs/CHANGELOG_CCIP.md:27]: could use some more detail on the usage of these new params, can be follow up to unblock

**Review by connorwstein — APPROVED** [review]

**Review by AnieeG — APPROVED** [review]
